src/hotspot/share/ci/ciMethodData.cpp

```diff
@@ -82,13 +82,20 @@ class PrepareExtraDataClosure : public CleanExtraDataClosure {
     return _safepoint_tracker.safepoint_state_changed();
   }
 
-  bool finish() {
+  bool finish(NoSafepointVerifier* no_safepoint) {
     if (_uncached_methods.length() == 0) {
       // Preparation finished iff all Methods* were already cached.
       return true;
     }
-    // Holding locks through safepoints is bad practice.
-    MutexUnlocker mu(_mdo->extra_data_lock());
+    // We are currently holding the extra_data_lock and ensuring
+    // no safepoint breaks the lock.
+    _mdo->check_extra_data_locked();
+    // We now want to cache some method data. This could cause
+    // a safepoint. We temporarily release the lock and allow
+    // safepoints, and revert that at the end of the scope:
+    MutexUnlocker mu(_mdo->extra_data_lock(), Mutex::_no_safepoint_check_flag);
```

Comment:
Why is this safe?

Reply:
We only do this in the `finish` method, where we hold no reference to any profiled-data anymore. We only really need to hold the lock during `clean_extra_data` and `is_live`. But after those are done, we can quickly release the lock so that we can call `get_method`.
Does that make sense to you? I'm not super happy with the general pattern here... I basically kept the old pattern.

I wonder, maybe there is a way to move the scope of the lock, such that we only need to lock inside of `clean_extra_data`, and do not hold it before we enter `clean_extra_data`. Do you think that would be preferable?

Reply:
@rwestrel I still believe this is safe. But maybe also ugly.

I looked into making the locking more fine-grained, so that we could avoid unlocking the lock temporarily.
The biggest problem is in `ciMethodData::load_remaining_extra_data`. Here we first (iteratively) clean, and then assume that we still hold the lock when we copy it for the `ciMethodData`. Hence, it seems the lock has to be held at this outer scope, but then temporarily unlocked to allow calls to `get_method` in `PrepareExtraDataClosure::finish`.

Reply:
Alternatives to make it prettier:
Make `prepare_metadata` lock, and pass out an object that holds that lock, i.e. widen the scope of the `MutexLocker`. Maybe this can be done with return-value-optimization? But I'm not sure this is a great idea. Another idea @chhagedorn and I thought about was having some Locker object that you can call lock/unlock on, repeatedly. But once the Locker goes out of scope, it checks if it is in the locked state, and only unlocks then. Or maybe it asserts that it is in the locked state, and then unlocks.

Because essencially we need to allow the retry-logic to unlock in between tries. But we also still need to access the `uncached_methods` array that is filled inside the locked region.

I'm not sure such a refactoring is worth it. Let me know what you think @rwestrel

Reply:
Given that logic existed, I think you can leave it as is but a comment that explains why it is safe would be useful.